File: Cracking The Coding Inverview - Book/01-arrays-and-strings/01-01-is-unique/isUnique.java
import java.util.Arrays;

public class isUnique {

    public static Boolean allUnique(String word) {

        // Represent all ASCII characters in a boolean array
        Boolean[] charSet = new Boolean[128];
        Arrays.fill(charSet, Boolean.FALSE);

        for (int i = 0; i < word.length(); i++) {
            int index = (int) word.charAt(i);
            if (charSet[index] == true) {
                return false;
            }
            charSet[index] = true;
        }
        return true;
    }

    public static void main(String[] args) {

        // Test if some words have all unique characters
        String[] words = {"Renzo", "Abigail", "Aitana", "Belon", "Giol"};
        System.out.println("Words with unique characters:");
        for (String word : words) {
            if (allUnique(word) == true) {
                System.out.println(word + " -> true");
            } else {
                System.out.println(word + " -> false");
            }
        }
    }
}

File: 01-arrays-and-strings/01-01-is-unique/isUnique.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

bool is_unique(char* word)
{
    int length = strlen(word);

    // Represent ASCII characters
    bool char_set[128] = { false };

    for (int i = 0; i < length; i++)
    {
        int val = word[i];
        if (char_set[val] == true)
        {
            return false;
        }
        char_set[val] = true;
    }
    return true;
}

int main(void)
{
    char* words[] = {"Renzo", "Abigail", "Aitana", "Belon", "Giol"};

    for (int i = 0; i < sizeof(words)/4; i++)
    {
        if (is_unique(words[i]) == true)
        {
            printf("%s -> true\n", words[i]);
        }
        else
        {
            printf("%s -> false\n", words[i]);
        }
    }
}
